### PR TITLE
[IMP] website,web_editor: target popup as a link

### PR DESCRIPTION
**Pull request by IT-Ideas**
Up to now, popup can be displayed after a delay or on exit.
This commit will allow the popup to be displayed on a click.

In order to do so, the user will have to:

	* Set the display property of the popup to OnClick
	* Copy the anchor (which is automatically displayed
	  in a toaster) into the url field of the link/button

task-2172312

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/static/src/snippets/s_popup/000.js [hunk 1]
```diff
@@ -37,21 +42,24 @@ const PopupWidget = publicWidget.Widget.extend({
...
-            this.timeout = setTimeout(() => this._showPopup(), delay);
+        if (this.display === 'afterDelay') {
+            this.timeout = setTimeout(() => this._showPopup(), this.delay);
+        } else if (this.display == 'onClick') {
+            $(document).ready(function () {
+                $("a[href$='#" + this.$el.attr('id') + "']").click(function (ev) {
+                    ev.preventDefault();
+                    this._showPopup();
+                }.bind(this));
+            }.bind(this));
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_popup/000.js:62]: You won't need this, Bootstrap has an automatic way to link a link to a modal. You should not have to modify this code but instead the link dialog code to add `data-toggle="modal"` when necessary.

Other reasons this is not ok:
- You'll never need document ready. Here you are in a public widget which is automatically instantiated on page load... and this instantiation occurs after DOM searches etc, which are made after the dom is ready. Also... our JS is lazy loaded and we start lazy loading once everything else is loaded ;)
- This is very rare to have to attach event handlers this way (most of the time, the widget `events` key should be used) and when you do need this, you have to unbind it somewhere (the destroy method). Otherwise, the event handler will persist when entering edit mode while it should not, or it will be duplicated as the widget are restarted when entering edit mode.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 2]
```diff
@@ -3772,8 +3772,167 @@ registry.SnippetSave = SnippetOptionWidget.extend({
...
+                options: {
+                    text: () => this._getAnchorLink(),
+                }
+            };
+            this._setupClipboardJSAndNotification(clipboardParams, () => {
+                const anchor = decodeURI(clipboardParams.options.text());
+                return {
+                    type: 'success',
+                    message: _.str.sprintf(_t("Anchor copied to clipboard<br>Link: %s"), anchor),
+                    buttons: [{text: _t("Edit"), click: () => this.openAnchorDialog(), primary: true}],
```
[the remark this answers is not in the document]
Comment (bvr-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3798]: I already send you this 2 small things to fix ;) I'll continue this review next week or after my vacation

[the remark this answers is not in the document]
Comment (IT-Ideas) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3798]: OK thanks, I indeed changed the function scope to private (added the underscore). It seems that the refactoring did not work :/

For the layout it is due to a bug introduced 3 to 4 weeks ago. I fixed it last week and it is now in master so I will just rebase. Thanks!

**Comment by bvr-odoo** [comment]
Use only dashes in the name of your branch ;)

**Comment by IT-Ideas** [comment]
> Use only dashes in the name of your branch ;)

This way I separated related branch, modules, description and employee. I was used to to that from the beginning :o) Did not know it was important and found it structured :o)

=> I will do so for future branches 👍

**Comment by bvr-odoo** [comment]
> > Use only dashes in the name of your branch ;)
> 
> This way I separated related branch, modules, description and employee. I was used to to that from the beginning :o) Did not know it was important and found it structured :o)
> 
> => I will do so for future branches


I thought it was in the guidelines but no .. so I think it's ok actually;)

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -3929,8 +3929,167 @@ registry.SnippetSave = SnippetOptionWidget.extend({
...
+ * Allows to copy the anchor to clipboard and optionally rename the anchor.
+ */
+const CopyAnchorToClipboardAndNotifyWidget = SnippetOptionWidget.extend({
+
+    /**
+     * @override
+     * @param {jQuery} $trigger the element that triggers the copy to clipboard and the notification
+     */
+    start: function ($trigger) {
+        if ($trigger) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3942]: Hum... how is this possible? `start` is a base method of the Widget class, it does not receive a parameter so it seems that the rest of this method does not ever do anything? Waiting your answer on that before continuing the review, maybe it is worth you check is still works and rebase it too :+1:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 3 above)
Comment (bvr-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3942]: @qsm-odoo It's possible because he made it possible ;) Look at the code in the start method of the snippet popup options. the parameter is added in the return => `return this._super.apply(this, dataAttributeDisplayOnClick || $(), arguments);`

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3942]: Yes he told me :wink: Needs to be changed to avoid that (with a simple write in a property in `this.`). As LTU works elsewhere, I told him you can take the branch again ;)

Comment (IT-Ideas) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3942]: Hello @bvr-odoo !
@qsb-odoo is right :o) I should have instead given a value to this.$trigger the call start of super.

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3942]: @IT-Ideas You left the team a week ago and you already forgot my trigram :cry: :sob: 
cc @JKE-be

Comment (IT-Ideas) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3942]: You know me 😉  I went too fast 🚗

Comment (JKE-be) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3942]: image

Comment (IT-Ideas) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3942]: 😄

Comment (JKE-be) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3942]: [Tu as deja oublié la website Team ?????????](https://www.youtube.com/watch?v=pZw2nRg1a-s&feature=youtu.be&t=64 "Tu m'as oublié")

Comment (IT-Ideas) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3942]: [https://www.youtube.com/watch?v=W_XEr3mH0Rc](url)

Comment (bvr-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: @qsm-odoo I made the change

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -4677,7 +4677,165 @@ registry.DynamicSvg = SnippetOptionWidget.extend({
...
+            };
+            this._setupClipboardJSAndNotification(clipboardParams, () => {
+                const anchor = decodeURI(clipboardParams.options.text());
+                return {
+                    type: 'success',
+                    message: _.str.sprintf(_t("Anchor copied to clipboard<br>Link: %s"), anchor),
+                    buttons: [{text: _t("Edit"), click: () => this._openAnchorDialog(), primary: true}],
+                };
+            });
+            return this._super.apply(this, arguments);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4704]: - `_super` has to be called even if there is no `$trigger`
- Use the ES6 form for that now: `this._super(...arguments)`

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 5]
```diff
@@ -4677,7 +4677,165 @@ registry.DynamicSvg = SnippetOptionWidget.extend({
     },
 });
 
+/**
+* Allows to copy the anchor to clipboard and optionally rename the anchor.
+*/
+const CopyAnchorToClipboardAndNotifyWidget = SnippetOptionWidget.extend({
+
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4684]: Remove that blank line

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 6]
```diff
@@ -4677,7 +4677,165 @@ registry.DynamicSvg = SnippetOptionWidget.extend({
...
+    /**
+    * @override
+    */
+    start: function () {
+        if (this.$trigger) { // the element that triggers the copy to clipboard and the notification
+            const clipboardParams = {
+                trigger: this.$trigger[0],
+                options: {
+                    text: () => this._getAnchorLink(),
+                }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4694]: Missing trailing comma

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 7]
```diff
@@ -4677,7 +4677,165 @@ registry.DynamicSvg = SnippetOptionWidget.extend({
     },
 });
 
+/**
+* Allows to copy the anchor to clipboard and optionally rename the anchor.
+*/
+const CopyAnchorToClipboardAndNotifyWidget = SnippetOptionWidget.extend({
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4683]: Why was this moved to web_editor? Keep it in website, especially since it is only used there...

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 8]
```diff
@@ -4677,7 +4677,165 @@ registry.DynamicSvg = SnippetOptionWidget.extend({
...
+     * Creates a safe id/anchor from text.
+     *
+     * @private
+     * @param {string} text
+     * @returns {string}
+     */
+    _text2Anchor: function (text) {
+        return encodeURI(text.trim().replace(/\s+/g, '-'));
+    },
+
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4833]: Remove this blank line

File: addons/website/static/src/snippets/s_popup/000.js [hunk 9]
```diff
@@ -39,21 +45,27 @@ const PopupWidget = publicWidget.Widget.extend({
...
             }
             this.$('.modal').removeClass('s_popup_middle').addClass('s_popup_bottom');
         }
 
-        if (display === 'afterDelay') {
-            this.timeout = setTimeout(() => this._showPopup(), delay);
+        if (this.display === 'afterDelay') {
+            this.timeout = setTimeout(() => this._showPopup(), this.delay);
+        } else if (this.display == 'onClick') {
+            let anchor = $.escapeSelector(this.$target.find('.modal').attr('id'));
```
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_popup/000.js:60]: Yes :+1:

File: addons/website/static/src/snippets/s_popup/000.js [hunk 10]
```diff
@@ -39,21 +45,27 @@ const PopupWidget = publicWidget.Widget.extend({
...
-        const $main = this.$target.find('.modal');
-
-        let display = $main.data('display');
-        let delay = $main.data('showAfter');
 
         if (config.device.isMobile) {
-            if (display === 'onExit') {
-                display = 'afterDelay';
-                delay = 5000;
+            if (this.display === 'mouseExit') {
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_popup/000.js:50]: - We cannot rename that... what about compatibility? EDIT: I see 'mouseExit' was introduced this way, so it seems right... but does not that mean this is a bug fix for stable versions?
- Also: can't this be moved to the start method as well? Feels strange to have moved the definition of this.display there but not this part

File: addons/website/static/src/snippets/s_popup/000.js [hunk 11]
```diff
@@ -39,21 +45,27 @@ const PopupWidget = publicWidget.Widget.extend({
...
+        } else if (this.display === 'onClick') {
+            let anchor = $.escapeSelector(this.$target.find('.modal').attr('id'));
+
+            const classesToAdd = {
+                'data-toggle': this.editableMode ? null : 'modal',
+                'data-target': this.editableMode ? null : '#' + anchor,
+            };
+            const $linkSelector = "a[href*='" + anchor + "']";
+            const $link = $($linkSelector);
+            $link.attr(classesToAdd);
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_popup/000.js:68]: I am not sure to understand this... should not this be the role of the option and not the public widget ? I think this widget should not be enabled in edit mode (like before).

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_popup/000.js (hunk 10 above)
Comment (bvr-odoo) [reply addons/website/static/src/snippets/s_popup/000.js:50]: - Yes it's actually a bug in stable, I add it in the "test website" task right now

File: addons/website/static/src/snippets/s_popup/000.js (hunk 11 above)
Comment (bvr-odoo) [reply addons/website/static/src/snippets/s_popup/000.js:68]: @IT-Ideas Hi ;) Could you explain why you made it like that please ?

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 7 above)
Comment (bvr-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:4683]: @IT-Ideas do you remember why you moved that here? can I put it back in website?

Comment (IT-Ideas) [reply addons/web_editor/static/src/js/editor/snippets.options.js:4683]: Hey! @bvr-odoo , I guess because I did not know about the snippet architecture and design at that time...

File: addons/website/static/src/snippets/s_popup/000.js (hunk 11 above)
Comment (IT-Ideas) [reply addons/website/static/src/snippets/s_popup/000.js:68]: Hey! @bvr-odoo , I guess because I did not know about the snippet architecture and design at that time...

Comment (bvr-odoo) [reply addons/website/static/src/snippets/s_popup/000.js:68]: @qsm-odoo Are you agree that it should be done with the link dialog (on save) ? Checking if the link is a modal and if yes adding the attributes

Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_popup/000.js:68]: Yes indeed ;)

**Comment by bvr-odoo** [comment]
I can't find a way to deactivate data-toggle = 'modal' in edit mode..
I tried stuff like that:
```
        $(document).on('click', '[data-toggle="modal"]', function (ev) {
            ev.preventDefault();
        });
```
but the modal is still open

**Comment by qsm-odoo** [comment]
> I can't find a way to deactivate data-toggle = 'modal' in edit mode..
> I tried stuff like that ( in the start of the editor.js file of web_editor):
> 
> ```
>         $(document).on('click', '[data-toggle="modal"]', function (ev) {
>             ev.preventDefault();
>         });
> ```
> 
> but the modal is still open

@bvr-odoo Did you try what we spoke about? https://getbootstrap.com/docs/4.6/components/modal/#modaldispose

**Comment by bvr-odoo** [comment]
> > I can't find a way to deactivate data-toggle = 'modal' in edit mode..
> > I tried stuff like that ( in the start of the editor.js file of web_editor):
> > ```
> >         $(document).on('click', '[data-toggle="modal"]', function (ev) {
> >             ev.preventDefault();
> >         });
> > ```
> > 
> > 
> > but the modal is still open
> 
> @bvr-odoo Did you try what we spoke about? https://getbootstrap.com/docs/4.6/components/modal/#modaldispose

yes but I did not succes to use it, however I finally have the solution to do it with stopPropagation()

**Comment by bvr-odoo** [comment]
@qsm-odoo 
I do not have the rights to change the state of the PR or to ask for a review.. So I ask here ;)

File: addons/web_editor/static/lib/summernote/src/js/module/Editor.js [hunk 12]
```diff
@@ -648,11 +650,18 @@ define([
...
         } else {
           $(anchor).removeAttr('target');
         }
+        if (isModalTarget) {
+          $(anchor)[0].dataset.toggle = 'modal';
+          $(anchor)[0].dataset.target = selector;
+        } else if ($(anchor)[0].dataset.toggle = 'modal') {
+          delete $(anchor)[0].dataset.toggle;
+          delete $(anchor)[0].dataset.target;
+        }
```
Comment (qsm-odoo) [inline addons/web_editor/static/lib/summernote/src/js/module/Editor.js:664]: Modification of the lib file ... this can't be good :/ Especially since summernote will be dropped by the new editor. Why would we need this? What is done here seems really the job of the link dialog

File: addons/website/views/snippets/snippets.xml [hunk 13]
```diff
@@ -934,12 +934,22 @@
     <!-- Anchor Name -->
     <div data-js="anchor"
         data-selector=":not(p).oe_structure > *, :not(p)[data-oe-type=html] > *"
-        data-exclude=".modal *, .oe_structure .oe_structure *, [data-oe-type=html] .oe_structure *">
+        data-exclude=".modal *, .oe_structure .oe_structure *, [data-oe-type=html] .oe_structure *, .s_popup, .o_newsletter_popup">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:937]: Bug fix?

File: addons/website/static/src/snippets/s_popup/options.js [hunk 14]
```diff
@@ -87,6 +87,15 @@ options.registry.SnippetPopup = options.Class.extend({
         const color = widgetValue ? 'var(--black-50)' : '';
         this.$target[0].style.setProperty('background-color', color, 'important');
     },
+    /**
+     * @see this.selectClass for parameters
+     */
+    addAnchor(previewMode, widgetValue, params) {
+        if (widgetValue) {
+            const widget = this._requestUserValueWidgets('anchor_opt')[0];
+            widget.el.click(); // TODO: to find why widget.enable() doesnt work here
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_popup/options.js:96]: TODO? Should be easy enough to debug I guess?

File: addons/website/views/snippets/snippets.xml [hunk 15]
```diff
@@ -934,12 +934,22 @@
...
         <we-button class="fa fa-fw fa-link o_we_link"
                    title="Create a link to target this section"
                    data-no-preview="true"/>
     </div>
 
+    <div data-js="anchor"
+        data-selector=".s_popup, .o_newsletter_popup"
+        data-target=".modal">
+        <div class="d-none">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:946]: I'd prefer a `ModalAnchor` JS option as an extension of the base anchor option and with `_computeVisiblity` overridden to not be shown, with an explanation in there.

Actually, even better: not create an element you don't use: you can use the notify system. Instead of finding the widget and simulate a click on it, trigger_up the "option_update" event and ask the anchor option to do the job

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/views/snippets/snippets.xml (hunk 13 above)
Comment (bvr-odoo) [reply addons/website/views/snippets/snippets.xml:937]: indeed.. fix here --> https://github.com/odoo/odoo/pull/65055

File: addons/website/views/snippets/snippets.xml (hunk 15 above)
Comment (bvr-odoo) [reply addons/website/views/snippets/snippets.xml:946]: With ClipboardJS, it's not possible de copy the link in the clipboard without using a click event (see [here](https://stackoverflow.com/questions/43235239/copy-text-to-clipboard-using-clipboard-js-without-button)
So are you okay if I don't use "option_update" and use your first solution

File: addons/website/static/src/snippets/s_popup/options.js (hunk 14 above)
Comment (bvr-odoo) [reply addons/website/static/src/snippets/s_popup/options.js:96]: no longer needed

File: addons/web_editor/static/lib/summernote/src/js/module/Editor.js (hunk 12 above)
Comment (bvr-odoo) [reply addons/web_editor/static/lib/summernote/src/js/module/Editor.js:664]: I haven't found a way to move 'data-attribute' (data-modal, data-target) from the link dialog to the editor. The only data that is handled are style, href and classes. I don't see how i can do it in an other way

Comment (qsm-odoo) [reply addons/web_editor/static/lib/summernote/src/js/module/Editor.js:664]: I don't understand, is not the link dialog in charge of changing href & co? Why can't it change data attributes too?

Comment (bvr-odoo) [reply addons/web_editor/static/lib/summernote/src/js/module/Editor.js:664]: As I understand, the link dialog pick up the information from the modal fields and give those information to the editor...
And then the editor build the link

Comment (qsm-odoo) [reply addons/web_editor/static/lib/summernote/src/js/module/Editor.js:664]: Dammit, what a mess :grimacing: Ok then but don't modify the lib file, you can override that `createLink` method in our own files for this case instead of modifying it directly in the lib file. Just like we override `showLinkDialog`.

Also, it would be great if you can add a test tour which drops a modal, create a link to that modal, save, click on the link and checks the modal is opened. We'll need better testing tools in the future but we should start making better test coverage especially for this kind of things that will simply be lost by the editor team if we don't make a test for it :wink:

**Comment by IT-Ideas** [comment]
Will be done in a new PR by @bvr-odoo
